File: js/linked_list.js
function Node(value){
    this.value = value
    this.next = null
}

function LinkedList(){
    this.length = 0;
    this.head = null;
    this.tail = null;
}

LinkedList.prototype.append = function (value) {
    
// debugger;
    var node = this.head // set Node to head

    if (node === null){
        console.log("empty")
        this.head = new Node(value)
        this.tail = this.head

    }else{
        newNode = new Node(value)
        this.tail.next = newNode
        this.tail = newNode

    }

    this.length ++
    console.log("append func ", node)
    console.log(this.length)
}

// LinkedList.prototype.appendMany = function(arr){

//     for(var i=0; i<arr.length; i++){

//     }
// }

LinkedList.prototype.insert = function (value, index){

    var length = this.length,
        currentNode = this.head,
        previousNode = null,
        position = 0
        newNode = new Node(value)

    //if index is greater than size or less than 0
    if(index > this.length || index < 0){
        return false
    }
    //if index is 0, insert at HEAD
    if(index === 0){
        var oldHead = this.head
        this.head = newNode

        this.head.next = oldHead

        this.length += 1

        return // exit
    }

    while(position < index){//why does this insert in current location?
        // debugger;
        var previous = currentNode,
            currentNode = currentNode.next
        console.log(currentNode)

        position += 1
    }

    
    previous.next = newNode
    newNode.next = currentNode

    this.length += 1;
    console.log("insert func ", newNode, this.length)
    console.log("insert func ", this.head, this.length)


}

LinkedList.prototype.remove = function(index){

    var length = this.length,
        currentNode = this.head,
        previousNode = null,
        nodeToDelete = null,
        position = 0
        // newNode = new Node(value)

    //if index is greater than size or less than 0
    if(index > this.length || index < 0){
        return false
    }
        
    //if index is 0, delete HEAD
    if(index === 0){

        var oldHead = this.head
        this.head = currentNode.next
        oldHead.next = null // remove node

        this.length -= 1

        return //exit
    }

    while(position < index){
    // while(currentNode.next !== null){
        // debugger
        previousNode = currentNode
        currentNode = currentNode.next
        // nodeToDelete = previousNode.next
        // currentNode = nodeToDelete.next

        position += 1

    }

    previousNode.next = currentNode.next
    console.log("list after node removed ", this.head)

}

LinkedList.prototype.printForward = function(){

    function printHelper(currentNode){

        if (currentNode === null) { return }
        
        printHelper(currentNode.next)
        console.log("print forward ", currentNode)

    }

    printHelper(this.head)

}

LinkedList.prototype.printBackward = function(){
  reverse(this.head);
}

function reverse(node) {
    var temp = new LinkedList()
  if(node === null) {
    return;
  }
    reverse(node.next);
    temp.append(node)
    console.log(node);
}


var testCase = new LinkedList();
testCase.append(10);
testCase.append(30);
testCase.append(20);
testCase.append(40);
// debugger;
testCase.insert(11, 0)//insert at HEAD
testCase.insert(11, -1)//insert at out of bound index
// debugger
testCase.insert(2, 3)//insert at given index
// testCase.append(3);
testCase.remove(0);
testCase.remove(2);

console.log('testCase', testCase)
testCase.printForward()
// testCase.printBackward()

File: python/linked_list.py
class Node(object):
   
   def __init__(self, value):
      self.value = value
      self.next = None


class LinkedList(object):
   
   def __init__(self):
      self.length = 0
      self.head = None
      self.tail = None


   def append(self, value):

      node = self.head

      if node == None:
         print ("empty node")
         self.head = Node(value)
         print(self.head.__dict__)
         # print(Node.__dict__)

         self.tail = self.head

         # return self.head
      else:
         print ("appending node " , value)
         new_node = Node(value)
         print("tails ", self.tail.__dict__)
         self.tail.next = new_node
         print("add new tail ", self.tail.__dict__)
         self.tail = new_node
         
         # return self.head

   
   def print_nodes(self):
      list1 = LinkedList()

      if list1.next == None:
         return
      else:
         print(self.head)


   def insert(self, parameter_list):
      
      for x in parameter_list:
         print (x)
         result = Node(x)
      
      return result

# node1 = Node(3)
# node2 = Node(23)
# node3 = Node(24)

list1 = LinkedList()


# ans = list1.insert([34,5,67,33])
ans = list1.append(35)
ans1 = list1.append(45)

# print(node1.__dict__)

# print(ans.__dict__)
# print("ans1.value ", ans1.value)
print(ans.__dict__)
print (ans1.__dict__)

LinkedList.print_nodes()
